Assets/Shopify/UIToolkit/Components/RemoteImageLoader.cs

```diff
@@ -0,0 +1,76 @@
+namespace Shopify.UIToolkit {
+    using UnityEngine;
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+
+    public interface ImageLoader {
+        string GetError();
+
+        Texture2D GetTexture();
+
+        IEnumerator Load(string url);
+    }
+
+    public class UnityImageLoader : ImageLoader {
+        private WWW _www;
+
+        public string GetError() {
+            return _www.error;
+        }
+
+        public Texture2D GetTexture() {
+            return _www.texture;
+        }
+
+        public IEnumerator Load(string url) {
+            _www = new WWW(url);
+            yield return _www;
+        }
+    }
+
+    /// <summary>
+    /// Completion callback that returns the downloaded texture and error description from a load operation.
+    /// </summary>
+    /// <param name="texture">Downloaded image as a Texture2D</param>
+    /// <param name="error">Error description</param>
+    public delegate void RemoteImageCompletionDelegate(Texture2D texture, string error);
+
+    /// <summary>
+    /// A behaviour that fetches a remote image (JPEG/PNG) from the web.
+    /// </summary>
+    public class RemoteImageLoader : MonoBehaviour {
```

Comment:
Same comment as before you broke this pr up into different pieces... if this is a service object for loading images it should not be a component... if its meant to be used alongside a UI.Image then we can abstract away getting the downloaded image an assigning it to a UI.Image class by including a UI.Image as a dependency.

My reasoning is that components describe some component of the total behaviour of an object in the game, if this is just a tool to load images remotely then creating a gameobject is extra work because the component doesnt need any features from the game object itself (for example, configuration in the editor or collaborating with neighboring components). Its best to keep it as a "code only" object as a result, because that is the intended way to interact with it.

If you need monobehaviour for the coroutine, another approach might be to have a static reference to a lazy instantiated GameObject, then you can call StartCoroutine on that game object to run a coroutine stored in your plain c# object